// igniteui-angular/esm5/lib/drop-down/drop-down.common.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @enum {number} */
var Navigate = {
    Up: -1,
    Down: 1,
};
export { Navigate };
Navigate[Navigate.Up] = 'Up';
Navigate[Navigate.Down] = 'Down';
/** @enum {string} */
var DropDownActionKey = {
    ESCAPE: 'escape',
    ENTER: 'enter',
    SPACE: 'space',
};
export { DropDownActionKey };
/**
 * Interface that encapsulates onSelection event arguments - old selection, new selection and cancel selection.
 * @export
 * @record
 */
export function ISelectionEventArgs() { }
if (false) {
    /** @type {?} */
    ISelectionEventArgs.prototype.oldSelection;
    /** @type {?} */
    ISelectionEventArgs.prototype.newSelection;
}
/**
 * Interface for an instance of IgxDropDownNavigationDirective
 * @export
 * @record
 */
export function IDropDownNavigationDirective() { }
if (false) {
    /** @type {?} */
    IDropDownNavigationDirective.prototype.target;
    /**
     * @param {?} event
     * @return {?}
     */
    IDropDownNavigationDirective.prototype.handleKeyDown = function (event) { };
    /**
     * @param {?=} event
     * @return {?}
     */
    IDropDownNavigationDirective.prototype.onArrowDownKeyDown = function (event) { };
    /**
     * @param {?=} event
     * @return {?}
     */
    IDropDownNavigationDirective.prototype.onArrowUpKeyDown = function (event) { };
    /**
     * @param {?=} event
     * @return {?}
     */
    IDropDownNavigationDirective.prototype.onEndKeyDown = function (event) { };
    /**
     * @param {?=} event
     * @return {?}
     */
    IDropDownNavigationDirective.prototype.onHomeKeyDown = function (event) { };
}
/** @type {?} */
export var IGX_DROPDOWN_BASE = 'IgxDropDownBaseToken';
/**
 * @hidden
 * @record
 */
export function IDropDownList() { }
if (false) {
    /** @type {?} */
    IDropDownList.prototype.onSelection;
    /** @type {?} */
    IDropDownList.prototype.width;
    /** @type {?} */
    IDropDownList.prototype.height;
    /** @type {?} */
    IDropDownList.prototype.id;
    /** @type {?} */
    IDropDownList.prototype.maxHeight;
    /** @type {?} */
    IDropDownList.prototype.collapsed;
    /** @type {?} */
    IDropDownList.prototype.items;
    /** @type {?} */
    IDropDownList.prototype.headers;
    /** @type {?} */
    IDropDownList.prototype.focusedItem;
    /**
     * @return {?}
     */
    IDropDownList.prototype.navigateFirst = function () { };
    /**
     * @return {?}
     */
    IDropDownList.prototype.navigateLast = function () { };
    /**
     * @return {?}
     */
    IDropDownList.prototype.navigateNext = function () { };
    /**
     * @return {?}
     */
    IDropDownList.prototype.navigatePrev = function () { };
    /**
     * @param {?} newIndex
     * @param {?=} direction
     * @return {?}
     */
    IDropDownList.prototype.navigateItem = function (newIndex, direction) { };
    /**
     * @param {?} key
     * @param {?=} event
     * @return {?}
     */
    IDropDownList.prototype.onItemActionKey = function (key, event) { };
}
/**
 * @hidden
 * @record
 */
export function IDropDownBase() { }
if (false) {
    /** @type {?} */
    IDropDownBase.prototype.selectedItem;
    /** @type {?} */
    IDropDownBase.prototype.onOpening;
    /** @type {?} */
    IDropDownBase.prototype.onOpened;
    /** @type {?} */
    IDropDownBase.prototype.onClosing;
    /** @type {?} */
    IDropDownBase.prototype.onClosed;
    /** @type {?|undefined} */
    IDropDownBase.prototype.allowItemsFocus;
    /**
     * @param {?} index
     * @return {?}
     */
    IDropDownBase.prototype.setSelectedItem = function (index) { };
    /**
     * @param {?} item
     * @param {?=} event
     * @return {?}
     */
    IDropDownBase.prototype.selectItem = function (item, event) { };
}
